amethyst_ui/src/button/mod.rs

```diff
@@ -0,0 +1,63 @@
+mod builder;
+mod system;
+
+pub use self::builder::{UiButtonBuilder, UiButtonBuilderResources};
+pub use self::system::UiButtonSystem;
+
+///! A clickable button.
+use amethyst_audio::SourceHandle;
+use amethyst_core::specs::prelude::{Component, DenseVecStorage};
+
+use amethyst_renderer::TextureHandle;
+
+/// A clickable button, this must be paired with a `UiImage`
+/// and this entity must have a child entity with a `UiText`.
+pub struct UiButton {
+    /// Default text color
+    normal_text_color: [f32; 4],
+    /// Default image
+    normal_image: TextureHandle,
+    /// Image used when the mouse hovers over this element
+    hover_image: Option<TextureHandle>,
+    /// Text color used when this button is hovered over
+    hover_text_color: Option<[f32; 4]>,
+    /// Image used when button is pressed
+    press_image: Option<TextureHandle>,
+    /// Text color used when this button is pressed
+    press_text_color: Option<[f32; 4]>,
+    /// Sound made when this button is hovered over
+    hover_sound: Option<SourceHandle>,
+    /// Sound made when this button is pressed.
+    press_sound: Option<SourceHandle>,
+}
+
+impl UiButton {
+    /// A constructor for this component.  It's recommended to use
+    /// either a prefab or `UiButtonBuilder` rather than this function
+    /// if possible.
+    pub fn new(
+        normal_text_color: [f32; 4],
+        normal_image: TextureHandle,
+        hover_image: Option<TextureHandle>,
+        hover_text_color: Option<[f32; 4]>,
+        press_image: Option<TextureHandle>,
+        press_text_color: Option<[f32; 4]>,
+        hover_sound: Option<SourceHandle>,
+        press_sound: Option<SourceHandle>,
```

Comment:
Instead of 1 builder per ui element,
you'll now have 1 builder, one struct and one system.

Reply:
prefab: If you only use composition of the basic ui elements and effect components, you can create presets (like a button) that you can reference after. When using 1 struct per element type, you are much more limited with what you can do (from the user perspective that is, because you need to either recode the button for your needs, or change it, which both require a fork)

Reply:
A composition based approach here gives redundancy in the prefab files.  You now have to specify the normal image and the normal text color twice.  That's what I'm trying to avoid here.

Reply:
And to be honest maybe we should have different builders for common UI elements.  They're convenience structures after all and requiring users to compose their own UI from such low level primitives introduces more cognitive overhead.

Reply:
You could make a wrapper that starts with your current builder and decomposes it into basic elements. Then in the prefabs you have the choice of going low level/composable or high level/configurable

Reply:
>And to be honest maybe we should have different builders for common UI elements. They're convenience structures after all and requiring users to compose their own UI from such low level primitives introduces more cognitive overhead.

Exactly. Internally (what interacts with systems) should only be low level components. What the user interfaces with should be either the low level or a higher level (previous comment)

Reply:
That's more or less what this PR does is it not?

Reply:
https://github.com/amethyst/amethyst/pull/798/files#diff-697d04ad45f512eac4d00b0353d63193R30

This is my issue. It is a internal system that uses a high level struct regrouping data that shouldn't be together.

Reply:
It's not possible to implement these features without that.  Where is button specific stuff supposed to be stored?  `UiImage` is too simple to permit that.

Reply:
>Component::OnHoverImage
System::OnHoverImageSystem
Component::OnHoverSound
System::OnHoverSound
struct UiButtonBuilder{
  hoverImage: Option<..>,
  .....
  ....
}
impl UiBuilder for UiButtonBuilder{
  fn build(self,world: &mut World) -> List<Entity>{
    world.create_entity().add(OnHoverImage).add(OnHoverSound)
  }
}

Reply:
It feels like more components for the sake of more components.  What do we gain from breaking it out like that?  Where would you use these other than on a button?  Keeping relevant data together helps with cognitive load and prevents `SystemData` tuples from getting obtusely large.  

I just worry it'd be `Anchor` and `Stretch` all over again.

Reply:
OnHoverImage
Used for: UiButton, UiCheckbox, UiDropdownList, UiTableCell, UiTab, UiPopup. So its pretty re-usable I think.

Now that I think about it, it might make more sense if they were not *that* separated.
Like anchor and stretch should have been directly into UiTransform.

Maybe OnUiActionSound(uievent, sound handle), OnUiActionImage(uievent, image)?

Reply:
Ok, that seems a bit more reasonable to me, but I do wonder about `OnUiActionText`.  In order to mimic current behavior it'd have to operate on the children of the `Entity` it's attached to, rather than the `Entity` it's currently attached to.

Reply:
Okay, so the fanciest that was actually doable was this:
```
OnUiAction<UiEventType,ComponentType>(target:Entity, |event:EventType, &mut component: ComponentType| -> ());
```
This works to edit components.

To perform side effects (playing sounds for example), you need to be more specific.
OnUiActionSound<UiEventType>(sound: AudioHandle)